File: minesweeper/java-minesweeper/src/main/java/it/chicio/minesweeper/field/parser/FieldRowContentChainParser.java
package it.chicio.minesweeper.field.parser;

public abstract class FieldRowContentChainParser implements FieldRowContentParser {
    private FieldRowContentChainParser nextFieldRowContentChainParser;

    public FieldsParsingStatus tryToParseRowAndUpdate(FieldsParsingStatus parsingStatus) {
        if (canParse(parsingStatus.currentRowContent)) {
            return parse(parsingStatus);
        } else if (hasAValidNextFieldRowParser()) {
            return nextFieldRowContentChainParser.tryToParseRowAndUpdate(parsingStatus);
        }
        return parsingStatus;
    }

    private boolean hasAValidNextFieldRowParser() {
        return nextFieldRowContentChainParser != null;
    }

    public void setNextFieldRowContentChainParser(FieldRowContentChainParser fieldRowContentChainParser) {
        this.nextFieldRowContentChainParser = fieldRowContentChainParser;
    }

    abstract boolean canParse(String row);

    abstract FieldsParsingStatus parse(FieldsParsingStatus fieldsParsingStatus);
}

File: minesweeper/kotlin-mockk-minesweeper/src/main/kotlin/it/chicio/minesweeper/field/parser/FieldRowContentChainParser.kt
package it.chicio.minesweeper.field.parser

abstract class FieldRowContentChainParser : FieldRowContentParser {
    private var nextFieldRowContentChainParser: FieldRowContentChainParser? = null
    override fun tryToParseRowAndUpdate(parsingStatus: FieldsParsingStatus): FieldsParsingStatus =
            when {
                canParse(parsingStatus.currentRowContent) -> parse(parsingStatus)
                else -> nextFieldRowContentChainParser?.tryToParseRowAndUpdate(parsingStatus) ?: parsingStatus
            }

    fun setNextFieldRowContentChainParser(fieldRowContentChainParser: FieldRowContentChainParser?) {
        nextFieldRowContentChainParser = fieldRowContentChainParser
    }

    abstract fun canParse(row: String?): Boolean
    abstract fun parse(fieldsParsingStatus: FieldsParsingStatus): FieldsParsingStatus
}